### PR TITLE
feat(mcp): add GitHub apply patch tool

### PR DESCRIPTION
**Pull request by riatzukiza**
## Summary
- add a github.apply_patch MCP tool that commits unified diffs to GitHub branches using the createCommitOnBranch mutation
- register the tool in the MCP registry and document it alongside existing GitHub helpers
- cover the new tool with tests for file additions and edits via mocked GitHub responses

## Testing
- pnpm --filter @promethean/mcp test

------
https://chatgpt.com/codex/tasks/task_e_68df3f6e6fb08324923ea62f8f94fbdb

### CODE REVIEW

**Review by riatzukiza — COMMENTED** [review]
Quick review summary (couldn’t formally REQUEST_CHANGES because GitHub blocks that for PR authors):

**Blockers to address before merge**
1) Input schema: export the compiled zod object as `inputSchema` (not the raw shape).
2) Tests: add AVA coverage for deletions‑only, rejected cases (binary/rename/unsafe path), `git apply` conflict, and mixed multi‑file patches.
3) Path decoding: current unescape is brittle; prefer a Git‑derived source of truth (e.g., parse `git apply --numstat --summary`) or a parser that follows Git’s quoting rules.

**Strong suggestions**
4) Preflight `git --version` and error clearly if missing.
5) Mode changes/symlinks: document or explicitly error.
6) Document size limits; current approach builds all fileChanges in memory.
7) Consider `UnsafePatchError` distinct from `GitApplyError`.
8) Nit: lean into functional transforms (map/flatMap) in builders.

**Docs**
• Add a Limitations section (no renames/binaries/mode changes; needs Git on PATH; size limits; conflict behavior).
• Include a minimal unified diff example + tool payload; show failure example & error payload.

**Open questions**
• Is `branch` required? If optional, resolve default branch; if required, enforce in zod.
• Renames/binary diffs: intentional hard‑fail? Document it.

Ping me after pushing updates and I’ll re‑review.

File: packages/mcp/test/tools.github.apply-patch.test.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+  t.true(result.ok);
+  t.is(result.commitOid, "fed321");
+  t.is(result.additions, 1);
+  t.is(result.deletions, 0);
+  t.true(
+    calls.some((call) =>
+      call.url.startsWith(
+        "https://api.github.com/repos/octo/demo/contents/src/app.txt",
+      ),
```
Comment (riatzukiza) [inline packages/mcp/test/tools.github.apply-patch.test.ts:190]: @codex Please add AVA tests to cover critical failure and edge cases: deletions‑only patch, binary diff, rename diff, unsafe path, and a `git apply` conflict. These should assert the correct error class/message and confirm no commit mutation occurs.

**Comment by riatzukiza** [comment]
Nice addition. A couple of must-haves for safety and DX. I’ve left specific suggestions inline where possible; for anything else, @codex please implement.

**Summary**
- Enforce diff size limits (bytes and file count).
- Ensure idempotency or at least a preflight head check.
- Prefer minimal schemas in specs (`inputSchema` as shape).
- Avoid binary/rename diffs (you already error out — good).
- Return structured errors so clients can branch.

Inline comments follow.

**Comment by riatzukiza** [comment]
Follow-ups with more specific snippets. @codex please apply these minimal edits.

1) **Size guardrails** — prevent pathological diffs

```ts
// after parsing `files`
if (files.length > 200 || diff.length > 1_000_000) {
  return { ok: false, kind: "SizeLimit", detail: { files: files.length, bytes: diff.length } };
}
```

2) **No-op short-circuit** — avoid empty commits

```ts
if (additions.length === 0 && deletions.length === 0) {
  return { ok: true, commitOid: null, commitUrl: null, additions: 0, deletions: 0 };
}
```

3) **Structured errors** — map known failures deterministically

```ts
catch (error) {
  if (error instanceof GitApplyError) {
    const details = [error.stdout.trim(), error.stderr.trim()].filter(Boolean).join("\n");
    return { ok: false, kind: "GitApplyError", code: error.code ?? null, details };
  }
  // optionally:
  // if (isGraphQlError(error)) return { ok: false, kind: "GitHubGraphQLError", message: error.message };
  return { ok: false, kind: "Unknown", message: String((error as any)?.message ?? error) };
}
```

4) **Tests** — add one for each branch above (limit, noop, git-apply error).

**Comment by riatzukiza** [comment]
Edge cases + tests for diff parsing. @codex please implement the following to close gaps.

**1) Normalize line endings before planning**
```ts
const normalizeEol = (s: string) => s.replace(/
?/g, '
');
const diff = normalizeEol(args.diff);
```

**2) Respect .gitattributes end-of-line policy (optional)**
- If repo has `* text=auto` or file-specific eol rules, normalize both sides accordingly before hunk match.

**3) File mode / symlink guard**
```ts
if (files.some(f => f.mode != null || f.isSymlink)) {
  return { ok:false, kind:'UnsupportedFileMode' };
}
```

**4) Hunk-edge handling** (add/delete at start/end of file)
- Allow zero-context matches at BOF/EOF; don’t require one extra context line.

**5) Test matrix (AVA)**
Add table-driven tests covering:

| case | input | expect |
|---|---|---|
| add at BOF | `+` lines at file start | 1 addition |
| delete at EOF | trailing lines removed | 1 deletion |
| CRLF diff vs LF repo | `
` in diff, repo LF | matches + applies |
| whitespace-only change (ignored) | spaces only w/ `ignoreWhitespace` | noop |
| rename/binary | diff indicates rename/binary | `UnsupportedDiffKind` |
| size limit | >200 files or >1MB | `SizeLimit` |
| rate limited | mock 403 w/ `X-RateLimit-Remaining: 0` | `RateLimited` |

Use fixtures: tiny repo tree snapshots + sample unified diffs. Keep everything pure/mocked; no network.

**Comment by riatzukiza** [comment]
Security + DX polish for apply_patch. @codex please implement.

**Secret redaction in logs**
- If you log GraphQL/REST errors, scrub Authorization headers and tokens before emitting.

**Commit message trailers**
- Add a trailer like `mcp: apply_patch` and a one-line summary of files changed; helps audit.

**Precondition safety**
- Always send `expectedHeadOid`; on mismatch return `{ ok:false, kind:'Conflict', expectedHeadOid, actualHeadOid }`.

**Retry policy**
- Idempotent retries on 502/503 with jittered backoff, max 3 attempts.

**Golden diffs**
- Store tiny unified diffs under `__fixtures__/apply_patch/` and snapshot the parsed plan in tests.

These keep ops safe and improve incident forensics.
